### ISSUE TITLE
[RFC] List/Repeating Field Spec

### ISSUE THREAD

**Issue by Pinpickle** [bolt/bolt#3267]
**TLDR;** I would like to make list fields happen.

This is an attempt to bring together the discussions around this topic, propose a method of accomplishing it, and to re-open the concept up for discussion.

Related issues/discussions: #1420 #1843 #2261 

People have brought up implementing libraries such as [Sir Trevor](http://madebymany.github.io/sir-trevor-js/) but I think that we should roll our own implementation. The most obvious concern is that Sir Trevor looks, feels and operates completely different to how Bolt does. Yes with some CSS we can style it ourselves and with some Javascript we can put in all of the hooks but in the end you're putting an awful lot of effort to re-create something our application can already do. The current fields just need to be given some more superpowers and magic, and after that the sky is the limit. Leveraging what we've already got means that adding new field won't complicate things, and extensions that add new fields will be able to be leveraged as well.

Onto the design.

## Design

I am proposing two distinct kinds of lists here, a dead simple one and one that allows for more complexity. Funnily enough, I think the dead simple one will be more difficult to implement.

### Simple Lists

**Simple lists** allow you to flag a single field as repeatable, and give a limit to how many may be added. These are stored and accessed as arrays of whatever kind of data the original field was.

#### Defining

These will be defined by appending a `limit` value to a field definition in a contenttype. By default, this value is set to `1`, so a field cannot be repeated. Setting it to an integer above 1 will allow that many fields in the list, and setting it to `0` will allow for as many items as the machine will allow.

If some people are obsessed with semantics, we could make the keyword `none` the same as `0` to say *there is no limit*.

```yml
staff:
    singular_slug: employee
    fields:
        capabilities:
            type: text
            limit: 3 # Can only list a maximum of 3 capabilities
        pets:
            type: text
            limit: 0 # These staff can have as many pets as they want
        name:
            type: text
            limit: 1 # This is superfluous, limit is 1 by default but it means the field can't repeat
```

#### Accessing in templates

Easy!

```twig
{% for capability in employee.capabilities %}
    {{ capability }}
{% endfor %}
```

#### UI in backend

Forgive my 5 minute mockups

![image](https://cloud.githubusercontent.com/assets/3238878/6864431/fd046716-d4c5-11e4-8d1c-7a6349ebfc0a.png)

A simple extension of each field type where it is shrunk on the right side to make room for a delete and dragging handle, and a new field button at the top.

There's probably a use-case for a duplicate handle as well.

### Repeating Blocks

This is where more complexity comes in, but probably more ease in implementation. This allows you to specify one or more sets of fields, like you would a contenttype, to repeat again and again. In terms of storing data, there are useful implications but I'm excited for this feature to use as long-form content, breaking free of the bounds of CKEditor to make interesting pages.

Yes, heavily inspired by [Craft CMS' matrix content](https://buildwithcraft.com/features/matrix) and later by Sir Trevor.

#### Defining

This is an entirely new fieldtype, `list` that has an option called `items` 

```yml
pages:
    fields:
        content:
            type: list
            item_limit: 20 # This is optional
            items:
                image:
                    fields:
                        image:
                            type: image # Yikes thats a lot of indents
                blockquote:
                    label: Block Quote
                    fields:
                        quote:
                            type: html
                        citation:
                            type: text
```

It speaks for itself, but it does get a bit unweildy and things would get complicated if you want to allow for identical sets of list fields inside other list fields - could probably get by with some YAML inheritance but that isn't super obvious to everyone.

So I'm also proposing a method of defining blocks to use in `config.yml`

```yml
list_blocks:
    image:
        fields:
            image:
            type: image
    blockquote:
        label: Block Quote
        fields:
            quote:
                type: html
            citation:
                type: text
```

Then if you supply an array of strings in your list field definition, it will read from list blocks

```yml
pages:
    fields:
        content:
            type: list
            item_limit: 20 # This is optional
            items:
              - image
              - blockquote
```

It also makes it fairly straight forward to have say, a 'section' type that can have more list fields inside it

```yml
list_blocks:
    image:
        fields:
            image:
            type: image
    blockquote:
        label: Block Quote
        fields:
            quote:
                type: html
            citation:
                type: text
    section:
        fields:
            colour:
                type: text
            content:
                type: list
                items:
                  - image
                  - blockquote
```

Defining blocks in `config.yml` will also stop many identical, but not defined in the same place, contenttype arrays clogging up memory which is a bonus.

**Question:** Should we allow to define the blocks in two areas or just one (in config or in contenttypes)?

**Potential syntactic sugar** for repeating one kind of block only, use `item` instead of `items`

```yml
section:
    type: list
    item:
        image:
            type: image
        heading:
            type: text
```

#### Accessing in templates

Slightly more complex, but still easy

```twig
{% for block in content %}
    {% set type = block.getSlug() %} {# The variable `slug` is commonly used so we need to find a different way to access the definition of this #}
    {% if type == 'image' %}
        {{ whatever }}
    {% elseif type == 'blockquote' %}
        {{ yeahsure }}
    {% endif %}
{% endfor %}
```

#### UI in backend

![image](https://cloud.githubusercontent.com/assets/3238878/6864529/66745a84-d4c7-11e4-95cb-a2cb4fa35d4e.png)

Now there are a lot of ways to do this. I've opted for putting every block in a little block that you can drag around and collapse. I see a problem being that any of these fields could end up with quite a lot of fields in them depending on how customisable you want things to get.

One option could be to use modals. That way, if your block has a lot of fields they'll never get in the way and will be treated pretty well. Gets a bit iffy with lists inside lists so I'm not a fan.

There is definitely a use case for tabs here, or maybe one or more collapsible sets of fields in there. These are things that the feature doesn't need to launch with and will be good to gauge use cases for.

## Implementation

I'm proposing this feature land in Bolt 2.3, and I'm very happy to take that on.

The first "dead simple" kind is near impossible to reliably implement with our current storage/content layer. As the new system will expose the logic of every field, managing that will be far easier. I don't think, though, that this part should be in 2.3 but 2.4 rather, a release after the new storage if (planned to be) out. This is just a simpler and more efficient version of the list field type so I would count it as a lower priority.

The second kind, which I am more excited about, shouldn't be too difficult to implement. The template refactoring in #3229 makes it much easier to store array-like fields based on other field types. 

### JS Refactoring

Original issue: #2781 
I've updated that issue and will not clog this with even more content

### Storage

I know nothing, [leave it to the experts](https://github.com/bolt/bolt/wiki/Bolt-2.x-Refactoring).

~~I'm in favour of storing it all as JSON encoded fields because I'm a big fan of a 1:1 ratio for contenttypes:tables and records:rows. It makes it so much easier for moving data around and it isn't a headache to look at. In my eyes, this only restricts methods of searching and retrieving data. I understand that the latter would be a concern for some people, but they might want to look towards using relationships or writing an extension for the more advanced functionality, or we define a standardised method of storing values in other tables.~~

~~I think if other values are going in other tables, it should be done in a transparent way where the developer is aware when its happening. Defaulting to it is not transparent.~~

### To extension or not to extension

I think this should be in core because it's a heavily requested feature and people - intentionally or subconsciously - work around the lack of features like this. Designers and developers want to empower their editors to create great things, and the lack of repeating content hurts that ability in quite a few ways. I'm mainly thinking of long form content but there are plenty more use cases than that.

It will also open things up for extension more. Someone could turn the repeating content field into a fully blown Mailchimp style wysiwyg page editor. I would consider that extension material, definitely.

I think the 'dead simple' kind of list field is pretty damn difficult to introduce as an extension and I don't know how you'd make all of the functionality it needs to changed that pluggable. 

Regardless, refactoring needs to happen either way. If this was an extension, there would most likely be areas that need to be made more extendable before even the list field type could land.

Jeepers, that got long. Enjoy!

**Comment by GawainLynch** [comment]
I'll come back to this with more :coffee: consumed… But :+1: on an epic RFC and a pretty cool idea!

**Comment by rossriley** [comment]
Ok, I'll jot down my thoughts here since this is a feature I've had my eye on getting sorted as part of the 2.x refactor.

I'm going to ignore the UI because I think how you've laid it out is pretty much how it would need to work.

### Part 1: Simple Repeatable fields
I'm going to suggest that we don't add this capability to every field but introduce a `repeater` field type, this could stay similarly simple to how you described, maybe something like this.

    staff:
        singular_slug: employee
        fields:
            capabilities:
                type: repeater
                field: text
                limit: 3
            pets:
                type: repeater
                field: text
                limit: 0
    ....

From an architecture point of view this will help because we can keep the model of 1 handler for each type of field. As part of the refactor we're also attempting to get this the same principle for joins such as relations and taxonomy so the `type` in config basically maps to a handler class, eg:

    type: text #src/Field/Type/Text
    type: float #src/Field/Type/Float
    # and then that follows:
    type: repeater #src/Field/Type/Repeater

### Part 2: Storage

I'm going to disagree with you on storing JSON in the database, that's not because it's necessarily a bad idea all the time, as you mentioned HTML is another kind of structured content we store. 

But I'm fairly certain that when we finally get this implementation of repeatable and composable fields correct then we will quickly see it being used as the primary tool for laying out content in Bolt sites.

You can think over any big project you may have done and suddenly there's a use for these fields, be it slider content, product variations, image galleries with captions, specification data etc. So fast forward a year after we ship this and we'll be in a situation where 30,40,50% maybe even more of an entire site is stored in JSON format in the database. 

If we get to that situation then not only do we have performance problems since we have none of the DB optimisation that comes with using non-blob fields. But we also have a problem when we want to search content or do more fine-grained queries. Think for example when you want to fetch all products with a 'yellow' variation available, in SQL you can write a query to check a single field and join, but in the JSON scenario we have to load all the relevant JSON for each product into a PHP structure before we can then do a search.

#### How would normalised data work?

I'll split this into HTML and DB structure, firstly the HTML structure would be fairly straightforward, a repeater field would signal to twig/js to use the following parts assuming our field is defined as the capabilities YML above:

    <input type="text" name="capabilities[]" value="">
    <input type="text" name="capabilities[]" value="">
    ... and more up to limit

The repeater field would receive a structure like this:
  
    'capabilities'=> [
        'capability value one',
        'capability value two'
    ]

And being a dedicated repeater handler it would know that it needs to normalise the data to the similar db structure described in the refactor doc: https://github.com/bolt/bolt/wiki/Bolt-2.x-Refactoring#field-database-table-problem

Equally the dedicated repeater field would know to do the reverse when it needs to fetch data from the db and return it to a standard array structure.

### Part 3: Composable Blocks

We had a discussion the other day on IRC about how you can currently do some nice stuff using YML anchors and references. (see: https://gist.github.com/rossriley/fdc9f545a8f122215334)

Again if we create a `block` type handler (or section as you used above) then we can allow a similar syntax to how you described, we could discuss whether we should create a separate config file for storing these definitions in, `blocks.yml`, `sections.yml` or maybe `fields.yml` but that's more aimed at keeping the YML files a bit more readable and not overly nested eg:

    ## in sections.yml or at the top of contenttypes.yml
    awesome_section: &awesome_section
        type: section
        limit: 3
        fields:
            title:
                type: text
                class: large
            image:
                type: image
            content:
                type: html
                height: 300px

    ## and then back in our normal contenttypes definition
    block-pages:
        name: Block Pages
        singular_name: Block Page
        fields:
            <<: *awesome_section
            extra_field:
                type: text

### TL;DR
Yes we should do this, no we shouldn't use JSON, we should create two new field types `repeater` and `section`.

**Comment by Pinpickle** [comment]
@rossriley Nice breakdown.

As I'm no data expert, not using JSON sounds like it is the way to go then. 

I have been quietly inspecting your repo every now and again to see what the new storage stuff will look like and I am loving it.

I'm not sure we should use yaml inheritance for re-using defined blocks. I've definitely found it really useful in contenttype config before, but I can see it causing sticky situations in the future. My main concern is not wanting to replicate data in the actual application.

These will be likely turned into contenttype objects at runtime. If we have all of the blocks defined, then we will only have a 1:1 contenttype:block ratio, whereas if you use yaml to replicate that data, it will be a many:1, which just doesn't sound right in my head, and it is a (tiny) extra memory overhead.

I see your point about the repeater field. Maybe a `limit` tag on a regular field type is just syntactic sugar for creating a repeater field in your spec? I just quite like the simplicity of definition.

**Comment by bobdenotter** [comment]
This is awesome! You guys are on fire! :fire::fire:

Also, "I mash my keyboard and software comes out" made me laugh out loud.

**Comment by jadwigo** [comment]
Wow.. this is a great start.

Also, I agree with @rossriley that json for storing this stuff will get us into trouble, searching records with serialized/json/blob content is hard.

I predict the next question we'll get is: "Can I use a repeater to link to another content type, any type. and mix it with simple text fields, images, movies, text..." If we have a solution for that bolt will probaly be the example for 99% of the other CMS out there.

And all of a sudden this will be the main method people will structure their sites. So please get a lot of feedback from UX people before releasing the first working version.

**Comment by CodeBrauer** [comment]
This is really [awesome](http://gph.is/VwOlaH). Great - this would made bolt complete for me.

**Comment by ellioseven** [comment]
+1

I want to move to Bolt but not being able to repeat sections of fields is a deal breaker. Say I have a company content type which has a "Contacts" field of phone numbers, email address, etc for sales, feedback, enquiries, etc. how could I accomplish this with Bolt? I don't think you can do this without some kind of repeatable fields? Forgive me if I'm wrong as I'm only trying to weight up wether its right for me.

**Comment by xiaohutai** [comment]
This topic is interesting :+1: 

@ellioseven If all you need is repeating multiple fields, then the [Gridviewfield extension](https://github.com/bobdenotter/gridviewfield) may be what you are looking for (note that the data is stored in json format).

**Comment by helman** [comment]
I really need this one. :+1: 

I just come to know Bolt for 10 days and love it. But this is the deal breaker.
When can we have it ? with in a month ? :dancer:

**Comment by GawainLynch** [comment]
@helman the foundation code with go into the 2.3 branch which should open soon, then it will be up to someone to implement (hopefully @Pinpickle) and then we're estimating a release in late July or early August of 2.3… but that might go longer.

**Comment by helman** [comment]
@GawainLynch nice to here that this will be rollout soom :D

@pinpickle This may be out of topic, i was thinking to make it more simple, instead of doing a block / complex inheritance, why not create a reference field type like relationship  option like we have right now (reference to contenttype), instead of only selecting the existing entry we able to add new entry (like sub entry) and than add limit option to limit the entry.
So we don't need to handle new type of data storage.
Sorry if this not makes sense. :D

**Comment by rossriley** [comment]
@helman you can already do the reference to a contenttype (and filter the results) via the select field type - see here for instructions. https://docs.bolt.cm/howto/select-field-type#populating-list-contenttype

**Comment by helman** [comment]
@rossriley yups, but that's only allow us to select from existing content (only do reference), we need to create and edit the content in separet view (page) before we able to select it. If there is a feature to create new or editing the existing reference content while adding content it will solve this feature. Isn't it ?
Sorry if my explenation confuse you.

**Comment by rossriley** [comment]
@helman yes you're right, misunderstood... in that case you'll need to wait for next version, it has been quite a hard feature to get implemented quickly because we've had to do a fair amount of refactoring of how Bolt handles storage internally, plus we're trying to do it in a way that won't break backwards compatibility too.

What you described could be implemented as an extension (in that a create button loads up a modal to create a new contenttype, then refereshes the select list) but that would come down to whether anyone has the time/inclination to develop this for 2.2 when a native solution is coming in 2.3.

**Comment by GawainLynch** [comment]
@rossriley is this closable now?

**Comment by rossriley** [comment]
yes
